### PR TITLE
New method canonicalize

### PR DESCRIPTION
**Pull request by weisslertal**
Added the method canonicalize which returns the canonical form of input regarding the base uri according to the RFC 1630

### CODE REVIEW

File: lib/uri/generic.rb [hunk 1]
```diff
@@ -1325,6 +1325,51 @@ def normalize!
...
+    # fully qualified path, server, path, file, search, quary.
+    # '.' and '..' are removed, where the latter also goes one dir up.
+    #
+
+    def canonicalize(input)
+      uri = 
+        if URI.parse(input).scheme
+          input
+        else #if it's not a qualified uri
+          if input[0,2].eql?('//') # if it starts with '//', set it as a host
```
Comment (nobu) [inline lib/uri/generic.rb]: `eql?` is for `Hash` look-up.
Use `==`, or `start_with?`.

File: lib/uri/generic.rb [hunk 2]
```diff
@@ -1325,6 +1325,51 @@ def normalize!
         set_host(self.host.downcase)
       end
     end
+
```
Comment (nobu) [inline lib/uri/generic.rb]: Remove trailing spaces.

File: lib/uri/generic.rb [hunk 3]
```diff
@@ -1325,6 +1325,51 @@ def normalize!
...
+          else
+            "#{File.dirname(self.to_s)}/#{input}" #if it doesn't start with any special characters, append it to the path
+          end
+        end
+    
+      uri = URI.parse(uri)
+      uri.normalize!
+      newpath = uri.path
+      unless newpath.eql?('/')
+        loop do
```
Comment (nobu) [inline lib/uri/generic.rb]: Removing `'..'` and `'.'` in this order seems buggy.
Consider `"foo/./../bar"`.

File: lib/uri/generic.rb [hunk 4]
```diff
@@ -1325,6 +1325,51 @@ def normalize!
...
+            "#{File.dirname(self.to_s)}/#{input}" #if it doesn't start with any special characters, append it to the path
+          end
+        end
+    
+      uri = URI.parse(uri)
+      uri.normalize!
+      newpath = uri.path
+      unless newpath.eql?('/')
+        loop do 
+          break unless newpath.gsub!(%r{([^/]+)/\.\./?}) { |match| $1 == '..' ? match : ''}
```
Comment (nobu) [inline lib/uri/generic.rb]: `loop do break unless COND` seems redundant.
It can be `nil while COND`.

File: lib/uri/generic.rb [hunk 5]
```diff
@@ -1325,6 +1325,51 @@ def normalize!
...
+    # Returns canonicalized URI.
+    #
+    #   require 'uri'
+    #
+    #   irb(main):929:0> URI("http://my.example.com/some_path").canonicalize "/dir/../dar/./dor/file.txt"
+    #   => #<URI::HTTP http://my.example.com/dar/dor/file.txt>
+    #
+
+    #
+    # Canonicalize means it supportes these types of input, according to the RFC:
```
Comment (stevensona) [inline lib/uri/generic.rb]: supportes is typo

**Comment by nurse** [comment]
URI#+ already behaves as you suggest
```
% irb
irb(main):001:0> u=URI("http://my.example.com/some_path") + "/dir/../dar/./dor/file.txt"
=> #<URI::HTTP http://my.example.com/dar/dor/file.txt>
irb(main):002:0> u.to_s
=> "http://my.example.com/dar/dor/file.txt"
```

**Pull request closed without merging**